### PR TITLE
 Fix assigning of error values to any constrained maps at runtime

### PR DESCRIPTION
**Pull request by Anoukh**
## Purpose
Fix the runtime check for inserting error values to an any constrained map, and nil values to non-nil constrained maps

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/14180

## Approach
Runtime always returns true if the map constraint is type any. This check was removed.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java [hunk 1]
```diff
@@ -4408,7 +4408,7 @@ private static boolean isValidMapInsertion(BType mapType, BValue value) {
         }
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java]: Not related to changes introduced in this PR, but we should check the constraint when the value is `null` right, whether it is a nil-able type?

The following doesn't fail atm, and allows an inherent type violating insertion!

```ballerina
import ballerina/io;

public function main(string... args) {
    map<string> m = { one: "one" };
    insertNilToMap(m);
    io:println(m);
}

function insertNilToMap(map<any> m) {
    m.two = ();
}
```

```cmd
$ ballerina run test.bal
{"one":"one", "two":()}
```

Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java]: Addressed with 2627ff2

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java [hunk 2]
```diff
@@ -4403,12 +4403,8 @@ private static void insertToMap(Strand ctx, BMap bMap, String fieldName, BValue
     }
 
     private static boolean isValidMapInsertion(BType mapType, BValue value) {
-        if (value == null) {
-            return true;
-        }
-
         BType constraintType = ((BMapType) mapType).getConstrainedType();
-        if (constraintType == BTypes.typeAny || constraintType.equals(value.getType())) {
+        if (value != null && constraintType.equals(value.getType())) {
```
Comment (MaryamZi) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BVM.java]: The first check in `checkIsType` (after checking whether the value is mutable) is `sourceType == targetType || sourceType.equals(targetType)`. So I guess we can remove this check and just call `checkIsType`, WDYT? We can do it separately maybe.

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectEquivalencyTest.java [hunk 3]
```diff
@@ -146,4 +146,13 @@ public void testObjectMemberOrder() {
         Assert.assertEquals(returns[0].stringValue(), "{age:45, name:\"Doe\", address:\"\"}");
         Assert.assertEquals(returns[1].stringValue(), "{age:35, name:\"John\", address:\"\"}");
     }
+
+
```
Comment (Anoukh) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectEquivalencyTest.java]:
```suggestion
```

**Pull request merged**